### PR TITLE
feat(eslint-plugin): [class-methods-use-this] add extension rule

### LINKED ISSUE #1103: [class-methods-use-this] Allow when required by interface
**Repro**

I'd like to request a `@typescript/class-methods-use-this` version of `class-methods-use-this`, which takes interfaces into account. If an interface requires a (public) method to be there, then the rule should allow it, even though it's not actually referencing `this`.

```JSON
{
  "rules": {
    "class-methods-use-this": "error"
  }
}
```

```ts
interface IStuff {
  doSomething(): void;
}

class Stuff implements IStuff {
  public doSomething(): void {
    // Do something without touching `this`.
  }
}
```

**Expected Result**
This to not flag `class-methods-use-this`.

**Actual Result**
Getting a `class-methods-use-this` violation on the `doSomething` method.

**Additional Info**

If this is impossible to do, I would also be happy with an `allowPublic` setting for that rule, that just skips checking public methods for usage of `this`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.4.0` |
| `@typescript-eslint/parser`        | `2.4.0` |
| `TypeScript`                       | `3.6.4` |
| `ESLint`                           | `6.5.1` |
| `node`                             | `10.15.0` |
| `npm`                              | `6.12.0` |

### PR DESCRIPTION
**Pull request by bradzacher**
## PR Checklist

- [x] Addresses an existing open issue: fixes #1103
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

This adds a standard AST fork of the base rule that can be configured to ignore one or both of the cases where a method has the `override` keyword, or the parent class implements an interface.

Based on #7260

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
@bradzacher is this still something you have time to work on?

**Comment by bradzacher** [comment]
I think it was almost done if memory serves me correctly. I'll have a look today.

File: packages/eslint-plugin/TSLINT_RULE_ALTERNATIVES.md [hunk 1]
```diff
@@ -185,7 +185,7 @@ It lists all TSLint rules along side rules from the ESLint ecosystem that are th
 | [`one-line`]                        | 🌟  | [`brace-style`][brace-style] or [Prettier]                                          |
 | [`one-variable-per-declaration`]    | 🌟  | [`one-var`][one-var]                                                                |
 | [`ordered-imports`]                 | 🌓  | [`import/order`]                                                                    |
-| [`prefer-function-over-method`]     | 🌟  | [`class-methods-use-this`][class-methods-use-this]                                  |
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/TSLINT_RULE_ALTERNATIVES.md:188]: [Non-actionable] Haha, I forgot about this file... good on you for remembering it!

File: packages/eslint-plugin/docs/rules/TEMPLATE.md [hunk 2]
```diff
@@ -1,6 +1,10 @@
+---
+description: '<Description from rule metadata here>'
+---
+
 > 🛑 This file is source code, not the primary documentation location! 🛑
 >
-> See **https://typescript-eslint.io/rules/your-rule-name** for documentation.
+> See **https://typescript-eslint.io/rules/RULE_NAME_REPLACEME** for documentation.
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/TEMPLATE.md:7]: unrelated?

Comment (bradzacher) [reply packages/eslint-plugin/docs/rules/TEMPLATE.md:7]: it was just me fleshing out the template
i noticed this was super easy to miss this so I made it more obvious!

File: packages/eslint-plugin/src/rules/class-methods-use-this.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+      }
+    }
+
+    return {
+      // function declarations have their own `this` context
+      FunctionDeclaration(): void {
+        pushContext();
+      },
+      'FunctionDeclaration:exit'(): void {
+        popContext();
```
[the remark this answers is not in the document]
Comment (bradzacher) [reply packages/eslint-plugin/src/rules/class-methods-use-this.ts:211]: I've got all the tests from the base rule - so either it's not covered by the base tests or code cov is wrong again. IDK

File: packages/eslint-plugin/docs/rules/TEMPLATE.md (hunk 2 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/TEMPLATE.md:7]: I like it :)
